File: medium/107_binary_tree_level_order_traversal_ii.rb
# 107. Binary Tree Level Order Traversal II
# Medium
#
# Given the root of a binary tree, return the bottom-up level order traversal
# of its nodes' values. (i.e., from left to right, level by level from leaf
# to root).

class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def level_order_bottom(root)
  level_order(root).reverse
end

def level_order(root)
  return [] if root.nil?

  res = []
  queue = [root]

  until queue.empty?
    level = []
    (0...queue.size).each do |_|
      cur = queue.shift
      level << cur.val
      queue << cur.left unless cur.left.nil?
      queue << cur.right unless cur.right.nil?
    end
    res << level
  end

  res
end

File: python/107_binary_tree_level_order_traversal_ii.py
"""
107. Binary Tree Level Order Traversal II
Medium

Given the root of a binary tree, return the bottom-up level order traversal
of its nodes' values. (i.e., from left to right, level by level from leaf
to root).
"""

from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrderBottom(self, root: Optional[TreeNode]) -> List[List[int]]:
        if root is None:
            return None

        res = []
        queue = deque([root])

        while len(queue) > 0:
            level = []
            for _ in range(len(queue)):
                curr = queue.popleft()

                level.append(curr.val)
                if curr.left is not None:
                    queue.append(curr.left)
                if curr.right is not None:
                    queue.append(curr.right)

            res.append(level)

        res.reverse()
        return res
